pyignite/monitoring.py

```diff
@@ -0,0 +1,457 @@
+    :ivar query_id: Query's id,
+    :ivar op_code: Operation's id,
+    :ivar op_name: Operation's name.
+    """
+    pass
+
+
+class QuerySuccessEvent(_QueryEvent):
+    """
+    Published when a client's query finished successfully.
+
+    :ivar host: Address of the node on which the query is executed,
+    :ivar port: Port number of the node on which the query is executed,
+    :ivar node_uuid: Node's uuid, string,
+    :ivar query_id: Query's id,
+    :ivar op_code: Operation's id,
+    :ivar op_name: Operation's name,
+    :ivar duration: Query's duration in milliseconds.
+    """
+    __slots__ = ('duration', )
+    duration: int
+
+    def __init__(self, host, port, node_uuid, query_id, op_code, op_name, duration, **kwargs):
+        super().__init__(host, port, node_uuid, query_id, op_code, op_name, duration=duration, **kwargs)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(host={self.host}, port={self.port}, " \
+               f"node_uuid={self.node_uuid}, query_id={self.query_id}, " \
+               f"op_code={self.op_code}, op_name={self.op_name}, duration={self.duration})"
+
+
+class QueryFailEvent(_QueryEvent):
+    """
+    Published when a client's query failed.
+
+    :ivar host: Address of the node on which the query is executed,
+    :ivar port: Port number of the node on which the query is executed,
+    :ivar node_uuid: Node's uuid, string,
+    :ivar query_id: Query's id,
+    :ivar op_code: Operation's id,
+    :ivar op_name: Operation's name,
+    :ivar duration: Query's duration in milliseconds,
+    :ivar error_msg: Error message.
+    """
+    __slots__ = ('duration', 'err_msg')
+    duration: int
+    err_msg: str
+
+    def __init__(self, host, port, node_uuid, query_id, op_code, op_name, duration, err, **kwargs):
+        super().__init__(host, port, node_uuid, query_id, op_code, op_name, duration=duration,
+                         err_msg=repr(err) if err else '', **kwargs)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(host={self.host}, port={self.port}, " \
+               f"node_uuid={self.node_uuid}, query_id={self.query_id}, op_code={self.op_code}, " \
+               f"op_name={self.op_name}, duration={self.duration}, err_msg={self.err_msg})"
+
+
+class QueryEventListener(_EventListener):
+    """
+    Base class for query event listeners.
+    """
+    def on_query_start(self, event: QueryStartEvent):
+        """
+        Handle query start event.
+
+        :param event: Instance of :class:`QueryStartEvent`.
+        """
+        pass
+
+    def on_query_success(self, event: QuerySuccessEvent):
+        """
+        Handle query success event.
+
+        :param event: Instance of :class:`QuerySuccessEvent`.
+        """
+        pass
+
+    def on_query_fail(self, event: QueryFailEvent):
+        """
+        Handle query fail event.
+
+        :param event: Instance of :class:`QueryFailEvent`.
+        """
+        pass
+
+
+class _EventListeners:
```

Comment:
Is it common practice in Python that there is a separate method for every event? Why won't we just add a `event_type` field to an event and `on_event` message to listener? With current approach we need to create new methods for every new event.

Reply:
>> With current approach we need to create new methods for every new event.
Yep, but it is much more convenient to user. Firstly, user should not create these methods by hand. Secondly,
it should not write a bunch if..elif..else code. Plus, python is a dynamic-typed language and there is not much help from IDE when you code such a thing. In current apporach these problems are eliminated.

Reply:
Here are the examples of PyCharm IDE help in this case:
1. Add new event handler
![Peek 2021-07-19 18-18](https://user-images.githubusercontent.com/6059537/126185744-d540cd74-9e0e-4692-9e6a-e2389dcc3e5f.gif)
2. Autocomplete fields of event
![Peek 2021-07-19 18-24](https://user-images.githubusercontent.com/6059537/126185809-25e093b1-be77-4656-8f25-5df37eb308b2.gif)